README.md
# drops-hull-extract-s3

[![CircleCI](https://circleci.com/gh/SMK1085/drops-hull-extract-s3.svg?style=svg)](https://circleci.com/gh/SMK1085/drops-hull-extract-s3)
[![codecov](https://codecov.io/gh/SMK1085/drops-hull-extract-s3/branch/master/graph/badge.svg)](https://codecov.io/gh/SMK1085/drops-hull-extract-s3)
[![License](https://img.shields.io/github/license/SMK1085/drops-hull-extract-s3.svg?style=flat-square)](https://github.com/SMK1085/drops-hull-extract-s3/blob/master/LICENSE) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://github.com/SMK1085/drops-hull-extract-s3#can-i-contribute-code) [![Code of conduct](https://img.shields.io/badge/code%20of-conduct-ff69b4.svg?style=flat-square)](https://github.com/SMK1085/drops-hull-extract-s3/blob/master/CODE_OF_CONDUCT.md)

A DROPS for requesting an extract from Hull and uploading it to AWS S3.

## Business Problem

This DROPS provides a sample solution to achieve the goal of periodically requesting extracts from the Hull platform and storing the result as CSV or JSON file in S3 to perform analysis or other tasks on the data.

## Solution Overview

This DROPS consists of two AWS Lambda functions which are written in Python. To deploy the sample solution easily, Terraform modules are available to provision all required AWS resources including the API Gateway, the S3 bucket and the necessary permissions, following the least privileged principle.

### Extract Requester Function

The purpose of this Lambda function is to request an extract from the Hull platform and is meant to be invoked via a scheduled event from AWS CloudWatch.
Since scheduled events do not carry any information, the configuration of what extract to ask for is done using environment variables of the Lambda function. The following variables are available:

| Variable | Description | Notes |
| -------- | ----------- | ----- |
|`HULL_CONNECTOR_ID`| The ID of the connector | Used to authenticate the extract request. Get it from any connector in your Hull organization |
|`HULL_CONNECTOR_SECRET`| The Secret of the connector | Used to authenticate the extract request. Get it from any connector in your Hull organizaton |
|`HULL_ORG_ID`| The ID of your Hull organization | This is the full ID of your Hull organization, e.g. `demo.hullapp.io`. Get it from any connector in your Hull organization |
|`HULL_EXPORT_CALLBACKURL`| The url of the Extract Processor Lambda Function | The url of the API Gateway Method to invoke the Extract Processor Function |
|`HULL_EXPORT_OBJECTTYPE`| The type of object to export. | Currently only `user` or `account` are supported.
|`HULL_EXPORT_FIELDS`| The fields to export in JSON encoded format. | Expects an array of fields, e.g. `["id", "external_id", "email", "hubspot/id"]`|
|`HULL_EXPORT_ESQUERY`| The ElasticSearch query to match objects against. | Expects a valid ElasticSearch query.|
|`HULL_LOG_LEVEL`| Optional. Defines the log level. | See section [Logging](#Logging) for details. |

### Extract Processor Function

The purpose of this Lambda function is to download the extracted file made available by the Hull platform and to handle the upload to S3.
This function requires a configured API Gateway method and access to the S3 bucket to upload the resulting files.

To configure the behavior, the following environment variables are available:

| Variable | Description | Notes |
| -------- | ----------- | ----- |
|`S3_BUCKET`| The name of the S3 bucket. | Note that the function needs to have permissions to at least upload files. |
|`S3_SINGLEFILE`| Optional. The name of the file to upload without extension.| You should set this variable only if you either want to overwrite the existing file (when `S3_DAILYFOLDER` is either not set or set to `False`) or if you want a consistent file name in your daily folder. |
|`S3_DAILYFOLDER`| Optional. Set it to `True`, `true` or `1` to enable a daily folder.| If this setting is enabled, the Lambda function will create a folder per day in the format YYYY-MM-DD, e.g. for Dec 26, 2019 it will create a folder named `2019-12-26` and add the resulting file to it. The day is determined by the UTC time when the function is invoked.|
|`S3_FILE_FORMAT`| Optional. Either `csv` or `json`, default is `csv`| Defines the file format to store data in. If not set it defaults to `csv`.|
|`HULL_EXPORT_FIELDS`| Optional. The fields to include in the S3 file in JSON encoded format. | Expects an array of fields, e.g. `["id", "external_id", "email", "hubspot/id"]`. If this variable is not set and the `S3_FILE_FORMAT` is set to `csv`, the first object in the extract will be used to determine the columns in the CSV file. This might lead to missing columns, so it is advised to keep the fields between the Requester and Processor functions in sync. If the `S3_FILE_FORMAT` is set to `json`, this parameter will be ignored. |
|`HULL_LOG_LEVEL`| Optional. Defines the log level. | See section [Logging](#Logging) for details. |

### Logging

Both Lambda functions come with logging, using the [logging library](https://docs.python.org/3/library/logging.html) from Python. If no log level is specified, the logger assumes a production environment and the log level will default to `ERROR`. To trace problems down, it is recommended to set the log level to `INFO` which will cause the functions to log every step with the relevant configuration. For further details about the output and how to retrieve the logs using CloudWatch, please refer to the [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/python-logging.html).

## Deployment using Terraform

### Requirements

Make sure that you have Terraform v0.12.0 or higher installed on your computer. You can download the latest version from the [Terraform download page](https://www.terraform.io/downloads.html).

### Examples

This repository contains various samples to get you started quickly with deploying everything to AWS. See the [Terraform examples folder](./tf/examples) for common quickstart scenarios:

| Scenario | Description |
| -------- | ----------- |
| [`extract-requester-only`](./tf/examples/extract-requester-only/README.md) | This scenario deploys a Lambda function which can request extracts from the Hull platform along with a CloudWatch scheduled event, so it can be executed on the defined schedule. |
| [`extract-processor-only`](./tf/examples/extract-processor-only/README.md) | This scenario deploys a Lambda function which can process extracts from the Hull platform along with the required API Gateway and S3 bucket. It also assigns IAM permissions to make all components work with each other. |
|[`extract-requester-w-processor`](./tf/examples/extract-requester-w-processor/README.md) | This scenario deploys a fully functioning combination of a extract requester Lambda along with a extract processor function, including CloudWatch scheduled event, API Gateway and S3 bucket. It also assigns IAM permissions to make all components work with each other. |
|[`multi-extract-requesters-w-processors`](./tf/examples/multi-extract-requesters-w-processors/README.md) | This scenario deploys a fully functioning combination of two extract requester Lambdas along with two extract processor functions, including CloudWatch scheduled events, API Gateway and S3 bucket. It also assigns IAM permissions to make all components work with each other. You would typically deploy this scenario if you want to process both, user and account extracts from Hull. |

Please refer to the respective READMEs of each scenario to learn how to deploy the scenario using Terraform. All code examples are configurable by adjusting the `.tfvars` files to your needs. All variables are described in the corresponding `variables.tf` file.

## Inspiration

- [terraform_aws_lambda_python](https://github.com/ruzin/terraform_aws_lambda_python)
- [blog-lambda-unit-testing-python](https://github.com/binxio/blog-lambda-unit-testing-python)
- [Introducing DROPS](https://medium.com/@smaschek85/introducing-drops-to-improve-dx-and-cx-for-saas-39a20f6e0f40)

## Credits

This project is authored and maintained by Sven Maschek ([@smaschek85](https://medium.com/@smaschek85) / [@svenmaschek](https://twitter.com/svenmaschek)).

Thank you to all [contributors](https://github.com/SMK1085/drops-hull-extract-s3/graphs/contributors).

## License

Open-source under [MIT License](https://github.com/SMK1085/drops-hull-extract-s3/blob/master/LICENSE).

## FAQ

### Uh oh, something went wrong!

Sorry about that. Please submit a bug report using the [GitHub issue tracker](https://github.com/SMK1085/drops-hull-extract-s3/issues).

### I wish something was different…

Keen to hear all ideas! Create an enhancement request using the [GitHub issue tracker](https://github.com/SMK1085/drops-hull-extract-s3/issues).

### Can I contribute code?

Yes please! See [CONTRIBUTING.md](./CONTRIBUTING.md) and [DEVELOPING.md](./DEVELOPING.md).

### My question isn't answered :(

Ask away using the [GitHub issue tracker](https://github.com/SMK1085/drops-hull-extract-s3/issues).


CHANGELOG.md
# CHANGLEOG

## v1.1.1

- Minor housekeeping items, added CI/CD and codecov badges

## v1.1.0

- Improved docs
- Added Terraform deployment scripts
- Added 4 deployment scenarios

## v1.0.0

- Initial Release without deployment scripts


tests/test_extract_processor.py
import os
from src.extract_processor import extract_processor
from tests import *
import unittest
from unittest.mock import Mock, patch
import requests
import responses
import pytest
import boto3
from moto import mock_s3
from datetime import datetime, timezone
import tempfile
import csv

messageMissingConfig = "Lambda function not properly configured. Missing environment variable '{0:s}'."

@mock_s3
class ExtractProcessorTests(unittest.TestCase):
    @pytest.fixture(autouse=True)
    def reset_env(self):
        os.environ.clear()

    def test_configerror_bucketname(self):
        os.environ["S3_BUCKET"] = ""

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert resp == { 'statusCode': 500, 'body': json.dumps(messageMissingConfig.format("S3_BUCKET")) }

    def test_configerror_s3fileformat(self):
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_FILE_FORMAT"] = "foo"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        error_msg = "Lambda function not properly configured. File format '{0:s}' is not one of the allowed values of 'csv' or 'json'."
        assert resp == { 'statusCode': 500, 'body': json.dumps(error_msg.format("foo")) }

    def test_invalidrequestbody(self):
        os.environ["S3_BUCKET"] = "hull-se-test"

        # Compose the event body
        body = { "foo": "bar" }
        
        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert resp == { 'statusCode': 200, 'body': json.dumps("Nothing to process.") }

    @responses.activate
    def test_extract_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Minimum env config
        os.environ["S3_BUCKET"] = "hull-se-test"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key="account_extract_5e032e118ee92e9586007425.csv")
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_singlefile_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Extended env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_SINGLEFILE"] = "my-single-file"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key="my-single-file.csv")
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_dayfolder_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Extended env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_DAILYFOLDER"] = "True"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        now = datetime.now(tz=timezone.utc)
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key=f"{now:%Y-%m-%d}/account_extract_5e032e118ee92e9586007425.csv")
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_dailysinglefile_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Extended env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_DAILYFOLDER"] = "True"
        os.environ["S3_SINGLEFILE"] = "my-daily-file"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        now = datetime.now(tz=timezone.utc)
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key=f"{now:%Y-%m-%d}/my-daily-file.csv")
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_singlefilewithfalsedaily_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Extended env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_DAILYFOLDER"] = "False"
        os.environ["S3_SINGLEFILE"] = "my-single-file2"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key="my-single-file2.csv")
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_bucketdoesntexist_fails(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/account_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.add(responses.GET, "https://somewhereinthecloud.local.host/extracts/account_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Extended env config
        os.environ["S3_BUCKET"] = "foo"
        os.environ["S3_SINGLEFILE"] = "my-single-file"

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/account_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert resp['statusCode'] == 500

    @responses.activate
    def test_userextract_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/user_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.reset()
            responses.add(responses.GET, "https://somewhereinthecloud.hull.net/extracts/user_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Minimum env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_SINGLEFILE"] = "my-userextract"
        os.environ['HULL_EXPORT_FIELDS'] = json.dumps(["id", "email", "name", "external_id", "segment_ids"])

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/user_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key="my-userextract.csv")
        
        with tempfile.TemporaryDirectory() as tmpdir:
            tmp_path = os.path.join(tmpdir, "my-userextract.csv")
            s3clnt.download_file("hull-se-test", "my-userextract.csv", tmp_path)
            print(tmp_path)
            with open(tmp_path, 'r') as csv_file:
                reader = csv.reader(csv_file)
                for i, line in enumerate(reader):
                    if i == 0:
                        assert line == ["id", "email", "name", "external_id", "segment_ids", "segment_names"]
            
            os.remove(tmp_path)
        
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }

    @responses.activate
    def test_jsonextract_is_successful(self):
        # Configure the mock to fake the get request
        download_content_path = os.path.join(os.path.dirname(__file__),"data/user_extract_download.json")
        with open(download_content_path, 'r') as download_file:
            res_body = download_file.read()
            responses.reset()
            responses.add(responses.GET, "https://somewhereinthecloud.hull.net/extracts/user_report/test.json", body=res_body)
        
        # Now let's fake S3 using moto
        conn = boto3.resource('s3', region_name='us-east-1')
        s3clnt = boto3.client('s3', region_name='us-east-1')
        # pylint: disable=no-member
        conn.create_bucket(Bucket="hull-se-test")
        
        # Minimum env config
        os.environ["S3_BUCKET"] = "hull-se-test"
        os.environ["S3_SINGLEFILE"] = "my-jsonextract"
        os.environ["S3_FILE_FORMAT"] = "json"
        os.environ['HULL_EXPORT_FIELDS'] = json.dumps(["id", "email", "name", "external_id", "segment_ids"])

        # Compose the event body
        body = None
        file_path = os.path.join(os.path.dirname(__file__),"data/user_extract.json")
        with open(file_path, "r") as outfile:
            reqObj = json.load(outfile)
            body = reqObj

        # Invoke the handler
        resp = extract_processor.handler(api_gateway_event(body), None)

        # Verify assertions
        assert None != s3clnt.get_object(Bucket="hull-se-test", Key="my-jsonextract.json")
        
        with tempfile.TemporaryDirectory() as tmpdir:
            tmp_path = os.path.join(tmpdir, "my-userextract.csv")
            s3clnt.download_file("hull-se-test", "my-jsonextract.json", tmp_path)
            with open(tmp_path, 'r') as json_file:
                data = json.load(json_file)
                assert 3 == len(data)
            
            os.remove(tmp_path)
        
        assert resp == {
            'statusCode': 200,
            'body': json.dumps("Successfully processed extract.")
        }


src/extract_requester/extract_requester.py
import os
import json
import requests
from requests.exceptions import HTTPError
from dotenv import load_dotenv
import logging

logger = logging.getLogger()

def __create_env_error_response(var_name: str):
    message = "Lambda function not properly configured. Missing environment variable '{0:s}'."
    logger.error(message.format(var_name))
    response = {
        'statusCode': 500,
        'body': json.dumps(message.format(var_name))
    }
    return response

def handler(event, context):
    # Load the environment variables
    load_dotenv()
    # Configure log level
    logger.setLevel(os.getenv("HULL_LOG_LEVEL", logging.ERROR))
    # Log invocation
    logger.info("Hull extract requester invoked with event: %s", event)

    # Parse the environment variables
    logger.info("Parsing configuration from environment variables...")

    HULL_ORG = os.getenv("HULL_ORG_ID")
    CONNECTOR_ID = os.getenv("HULL_CONNECTOR_ID")
    CONNECTOR_SECRET = os.getenv("HULL_CONNECTOR_SECRET")
    CALLBACK_URL = os.getenv("HULL_EXPORT_CALLBACKURL")
    HULL_OBJECTTYPE = os.getenv("HULL_EXPORT_OBJECTTYPE")
    HULL_FIELDS = os.getenv("HULL_EXPORT_FIELDS")
    HULL_ESQUERY = os.getenv("HULL_EXPORT_ESQUERY")

    logger.info("Loaded environment variables: %s", { 
        "HULL_ORG_ID": HULL_ORG,
        "HULL_CONNECTOR_ID": CONNECTOR_ID,
        "HULL_CONNECTOR_SECRET": CONNECTOR_SECRET,
        "HULL_EXPORT_CALLBACKURL": CALLBACK_URL,
        "HULL_EXPORT_OBJECTTYPE": HULL_OBJECTTYPE,
        "HULL_EXPORT_FIELDS": HULL_FIELDS,
        "HULL_EXPORT_ESQUERY": HULL_ESQUERY
    })

    # Perform pre-flight checks
    if not HULL_ORG:
        return __create_env_error_response("HULL_ORG_ID")
    
    if not CONNECTOR_ID:
        return __create_env_error_response("HULL_CONNECTOR_ID")
    
    if not CONNECTOR_SECRET:
        return __create_env_error_response("HULL_CONNECTOR_SECRET")
    
    if not CALLBACK_URL:
        return __create_env_error_response("HULL_EXPORT_CALLBACKURL")

    if not HULL_OBJECTTYPE:
        return __create_env_error_response("HULL_EXPORT_OBJECTTYPE")
    
    if not HULL_FIELDS:
        return __create_env_error_response("HULL_EXPORT_FIELDS")

    if not HULL_ESQUERY:
        return __create_env_error_response("HULL_EXPORT_ESQUERY")

    if HULL_OBJECTTYPE not in ["user", "account"]:
        message = "Lambda function not properly configured. Object type '{0:s}' is not one of the allowed values of 'user' or 'account'."
        logger.error(message.format(HULL_OBJECTTYPE))
        response = {
            'statusCode': 500,
            'body': json.dumps(message.format(HULL_OBJECTTYPE))
        }
        return response

    # All pre-flight checks have passed, so we are ready to request the extract
    try:
        export_url = f"https://{HULL_ORG}/api/v1/extract/{HULL_OBJECTTYPE}s"
        req_body = {
            "url": CALLBACK_URL,
            "format": "json",
            "fields": json.loads(HULL_FIELDS),
            "query": json.loads(HULL_ESQUERY)
        }
        req_headers = {
            "Accept": "application/json",
            "Hull-App-Id": CONNECTOR_ID,
            "Hull-Access-Token": CONNECTOR_SECRET
        }
        logger.info("Requesting extract from the Hull platform at url '%s' with headers %s and body %s", export_url, req_headers, req_body)
        response = requests.post(export_url, json=req_body, headers=req_headers)
        logger.info("Extract requested, finished with status: %s %s", response.status_code, response.reason)
        response.raise_for_status()

    except HTTPError as http_error:
        logger.error("An unexpected HTTP error occurred: %s", http_error)
        response = {
            'statusCode': http_error.response.status_code,
            'body': json.dumps(f"An unexpected HTTP error occurred: {http_error}")
        }
        return response
    except Exception as err:
        logger.error("An unexpected error occurred: %s", err)
        response = {
            'statusCode': 500,
            'body': json.dumps(f"An unexpected error occurred: {err}")
        }
        return response
    else:
        logger.info("Successully requested export from Hull. Result will be sent to '%s'", CALLBACK_URL)
        response = {
            'statusCode': 200,
            'body': json.dumps(f"Successully requested export from Hull. Result will be sent to '{CALLBACK_URL}'")
        }
        return response
